query/query.go

```diff
@@ -182,22 +179,23 @@ func (r *results) Rest() ([]Entry, error) {
 		}
 		es = append(es, e.Entry)
 	}
-	<-r.proc.Closed() // wait till the processing finishes.
+	<-r.Done() // wait till the processing finishes.
 	return es, nil
 }
 
-func (r *results) Process() goprocess.Process {
-	return r.proc
-}
-
-func (r *results) Close() error {
-	return r.proc.Close()
+func (r *results) Close() {
+	r.cancel()
+	<-r.closed
```

Comment:
close(r.closed) instead ? (to avoid multiple Done() callers from hanging).

Reply:
The call to `r.cancel()` tells the results processing goroutine to stop. Then, `<-r.closed` waits for it to finish stopping. The goroutine needs to do `close(r.closed)` on exit.

This is designed for multiple callers of `rClose()` to all wait for the processing goroutine to exit.

I think the missing piece is that `ResultsBuilder` needs to provide access to its `ctx` and `cancel` to allow it to be used outside `ResultsWithContext`. Or... `ResultsBuilder` should just go away.